Fixes Part of #8472: Migrate Search Service to Angular 8

SarahZLipeles opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #8472 .
2. This PR does the following: Migrates Search Service to Angular 8

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.

Commit:
migrate SearchService to Angular

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

Commit:
migrate SearchService tests from AngularJS

Commit:
rename selection details
return loadMoreData errorCallback to previous functionality

Commit:
rename SelectionDetails

Commit:
add success handler for executeSearchQuery

Commit:
add SearchService to angular-services

Commit:
remove fdescribe

Commit:
revert SearchService loadMoreData functionality

Commit:
add importAllAngularServices where needed

Commit:
remove eslint disable camelcase

Commit:
move http call into backend api service

Commit:
remove unneccessary $apply

Commit:
add SearchBackendApiService to angular-services

SarahZLipeles commented:
@sagangwee Tagging you as requested

srijanreddy98 commented on core/templates/services/search.service.ts:
```diff
@@ -16,215 +16,234 @@
  * @fileoverview search service for activityTilesInfinityGrid
```
```suggestion
 * @fileoverview Search service for activityTilesInfinityGrid.
```

SarahZLipeles replied:
done

srijanreddy98 commented on core/templates/services/search.service.ts:
```diff
@@ -16,215 +16,234 @@
-            }
-            _languageCodes += key;
-          }
+  private getSuffixForQuery(
```
We add `_` in front of private variables and methods. Can you change that here and also to the private stuff above.
```suggestion
  private _getSuffixForQuery(
```

SarahZLipeles replied:
done

srijanreddy98 commented on core/templates/services/search.service.ts:
```diff
@@ -16,215 +16,234 @@
-          }
+  private getSuffixForQuery(
+      selectedCategories:SelectionList,
+      selectedLanguageCodes:SelectionList) {
```
Can you add a return type here as well? I think this returns a string.

SarahZLipeles replied:
done

srijanreddy98 commented on core/templates/services/search.service.ts:
```diff
@@ -16,215 +16,234 @@
-    var hasReachedEndOfPage = function() {
-      return _searchCursor === null;
-    };
+  hasReachedEndOfPage():boolean {
```
Add a space after `:`. Also do this for other function here.
```suggestion
  hasReachedEndOfPage(): boolean {
```

SarahZLipeles replied:
done

srijanreddy98 commented on core/templates/services/search.service.ts:
```diff
@@ -16,215 +16,234 @@
-      get onInitialSearchResultsLoaded() {
-        return _initialSearchResultsLoadedEventEmitter;
+  loadMoreData(
+      successCallback:(SearchResponseData, boolean)=>void,
```
Space before and after `=>`
```suggestion
      successCallback: (SearchResponseData, boolean) => void,
```

SarahZLipeles replied:
done

srijanreddy98 commented on core/templates/services/search.service.ts:
```diff
@@ -16,215 +16,234 @@
+  */
+  updateSearchFieldsBasedOnUrlQuery(
+      urlComponent:string, selectionDetails:SelectionDetails): string {
+    var urlQuery = urlComponent.substring('?q='.length);
```
let/ const instead of var?

SarahZLipeles replied:
done

srijanreddy98 commented on core/templates/pages/library-page/search-results/activity-tiles-infinity-grid.component.ts:
```diff
@@ -40,6 +43,7 @@ angular.module('oppia').component('activityTilesInfinityGrid', {
                 data.activity_list));
             ctrl.endOfPageIsReached = endOfPageIsReached;
             ctrl.searchResultsAreLoading = false;
+            $rootScope.$apply();
```
Use `$rootScope.$applyAsync()` instead.

SarahZLipeles replied:
done

srijanreddy98 reviewed (changes requested):
@TheoLipeles took a pass. PTAL!

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

Commit:
rename getSearchresults to getSearchResults

Commit:
move SEARCH_DATA_URL into backend api service

Commit:
add tests for Search Backend Api Service

Commit:
fix syntax inconsistencies

SarahZLipeles commented:
The review above was sent by a vscode extension before I pushed the code, code should now be updated

SarahZLipeles commented:
@srijanreddy98 PTAL
I fixed the changes you specified and added a spec file for Search Backend Api Service

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

srijanreddy98 reviewed (approved):
LGTM! @kevintab95 @nithusha21 PTAL!

kevintab95 reviewed (approved):
LGTM, thanks @TheoLipeles! Just wanted to check, have you tested the search functionality on your local machine and confirmed that it works as expected?

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

SarahZLipeles commented:
> LGTM, thanks @TheoLipeles! Just wanted to check, have you tested the search functionality on your local machine and confirmed that it works as expected?

Yup

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

Commit:
Merge branch 'migrate-angular-search-service' of https://github.com/TheoLipeles/oppia into migrate-angular-search-service

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

nithusha21 commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -0,0 +1,64 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Service for user data.
```
wrong overview?

SarahZLipeles replied:
done

nithusha21 commented on core/templates/services/search.service.spec.ts:
```diff
@@ -16,87 +16,56 @@
  * @fileoverview Tests that average ratings are being computed correctly.
```
Wrong overview here too?

SarahZLipeles replied:
done

nithusha21 reviewed (commented):
Overall looks good... Just a few nits

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

Commit:
fix fileoverviews

SarahZLipeles commented:
@nithusha21 fixed the file overviews PTAL

nithusha21 commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -13,7 +13,7 @@
 
 /**
- * @fileoverview Service for user data.
+ * @fileoverview Service for searching for collections
```
Add a full stop at the end of the overview.

SarahZLipeles replied:
done, this was also still wrong,
It's a search service for explorations not collections

Commit:
fix fileoverview

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

DubeySandeep commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -0,0 +1,64 @@
+export class SearchBackendApiService {
+  constructor(private http: HttpClient) {}
+
+  async getSearchResults(searchQuery:string): Promise<SearchResponse> {
```
getSearchResultsAsync* (We are expecting to have Async suffix for async method.)

SarahZLipeles replied:
done

DubeySandeep commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -0,0 +1,64 @@
+export class SearchBackendApiService {
+  constructor(private http: HttpClient) {}
+
+  async getSearchResults(searchQuery:string): Promise<SearchResponse> {
```
Either change the function name to fetchExplorationSearchResult or change the file name to exploration-search-backend-api.service (As of now SearchBackendApiService.getSearchResults doesn't indicate what we are searching for.)

I'll prefer the change in function name so that we can have other search function here in the future if needed. (If you are changing the function name then do update the fileoverview to indicate that this service is for general search API.)

SarahZLipeles replied:
went with fetchExplorationSearchResultAsync
done

DubeySandeep commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -0,0 +1,64 @@
+}
+
+angular.module('oppia').factory(
+  'UserBackendApiService',
```
Why is it `UserBackendApiService`?

SarahZLipeles replied:
I used `UserBackendApiService` as a template while creating this,
fixed

DubeySandeep commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -0,0 +1,64 @@
+  summary: string;
+}
+
+export interface SelectionDetails {
```
is this interface used here? if not then any specific reason for defining it here? (In general, we define interface only for backend response, and for general frontend objects we use class/objects.)

/cc @srijanreddy98, I think we need to define possible places to use the interface.

SarahZLipeles replied:
these are used in search service, I moved them and changed them to classes
fixed

DubeySandeep commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -0,0 +1,64 @@
+
+export interface SelectionList {[key: string]: boolean}
+
+export interface FilterDetails {
```
Ditto*

SarahZLipeles replied:
these are used in search service, I moved them and changed them to classes
fixed

DubeySandeep commented on core/templates/services/search-backend-api.service.ts:
```diff
@@ -0,0 +1,64 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Service for searching for explorations.
```
The filename doesn't indicate that this service is only for exploration.

SarahZLipeles replied:
I changed it to be more generally for search queries
fixed

DubeySandeep commented on core/templates/services/search-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,55 @@
+      });
+      const req = httpTestingController.expectOne('/searchhandler/data');
+      expect(req.request.method).toEqual('GET');
+      req.flush(sampleSearchResponse);
```
Assert the response here*

SarahZLipeles replied:
done

DubeySandeep commented on core/templates/services/search.service.ts:
```diff
@@ -13,218 +13,238 @@
+
+    this._isCurrentlyFetchingResults = true;
+    this.numSearchesInProgress++;
+    this._searchBackendApiService.getSearchResults(queryUrl)
```
Should we move the url generation code in the backend.api.service and here pass getSearchUrlQueryString as a param? (As backend-api.service handling URL related stuff)

SarahZLipeles replied:
This currently isolates the backend url manipulation (adding `/searchhandler`) to the backend service api,
but the url manipulation in Search Service could go either way, it seems to be intended for both the browser url bar, and the backend query

DubeySandeep reviewed (changes requested):
@TheoLipeles, I took a quick pass on the* backend-api.service and left a few comments, PTAL!

Commit:
Merge remote-tracking branch 'upstream/develop' into migrate-angular-search-service

Commit:
change search backend api service
to be a service for general searches
to specify exploration search
to use proper downgrade name
remove unneeded interfaces

Commit:
export SelectionList and SelectionDetails from SearchService

SarahZLipeles commented:
Testing locally, it seems that the search service is slowed down when compared with the develop branch,
I'm looking into this

Commit:
add $rootScope.$applyAsync on initial results loaded

SarahZLipeles commented:
> Testing locally, it seems that the search service is slowed down when compared with the develop branch,
> I'm looking into this

Fixed

SarahZLipeles commented:
@DubeySandeep I fix the issues you listed and responded with a small design decision question, PTAL

DubeySandeep commented on core/templates/services/search.service.ts:
```diff
@@ -13,218 +13,259 @@
+  summary: string;
+}
+
+export class SelectionDetails {
```
@TheoLipeles, FYI, I'll talk to angular team about dict-class and review the PR by tomorrow.

DubeySandeep reviewed (approved):
LGTM!
